src/value_log.rs

```diff
@@ -0,0 +1,331 @@
+    dir.as_ref().join(format!("{:06}.vlog", fid))
+}
+
+struct Core {
+    /// `files_map` stores mapping from value log ID to WAL object.
+    /// TODO: use scheme like memtable to separate current vLog
+    /// and previous logs, so as to reduce usage of Mutex.
+    files_map: HashMap<u32, Arc<Mutex<Wal>>>,
+    /// maximum file ID opened
+    max_fid: u32,
+    files_to_delete: Vec<u32>,
+    num_entries_written: u32,
+}
+
+impl Core {
+    fn new() -> Self {
+        Self {
+            files_map: HashMap::new(),
+            max_fid: 0,
+            files_to_delete: vec![],
+            num_entries_written: 0,
+        }
+    }
+}
+
+/// ValueLog stores all value logs of an agatedb instance.
+pub struct ValueLog {
+    /// value log directory
+    dir_path: PathBuf,
+    /// value log file mapping, use `RwLock` to support concurrent read
+    core: Arc<RwLock<Core>>,
+    /// offset of next write
+    writeable_log_offset: AtomicU32,
+    opts: AgateOptions,
+}
+
+impl ValueLog {
+    /// Create value logs from agatedb options
+    pub fn new(opts: AgateOptions) -> Option<Self> {
+        if opts.in_memory {
+            None
+        } else {
+            Some(Self {
+                core: Arc::new(RwLock::new(Core::new())),
+                dir_path: opts.value_dir.clone(),
+                opts,
+                writeable_log_offset: AtomicU32::new(0),
+            })
+            // TODO: garbage collection
+            // TODO: discard stats
+        }
+    }
+
+    fn file_path(&self, fid: u32) -> PathBuf {
+        vlog_file_path(&self.dir_path, fid)
+    }
+
+    fn populate_files_map(&self) -> Result<()> {
+        // TODO: implement
+        Ok(())
+    }
+
+    fn create_vlog_file(&self) -> Result<(u32, Arc<Mutex<Wal>>)> {
+        let mut core = self.core.write().unwrap();
+        let fid = core.max_fid + 1;
+        let path = self.file_path(fid);
+        let wal = Wal::open(path, self.opts.clone())?;
+        // TODO: only create new files
+        let wal = Arc::new(Mutex::new(wal));
+        assert!(core.files_map.insert(fid, wal.clone()).is_none());
+        assert!(core.max_fid < fid);
+        core.max_fid = fid;
+        // TODO: add vlog header
+        self.writeable_log_offset
+            .store(0, std::sync::atomic::Ordering::SeqCst);
+        core.num_entries_written = 0;
+        Ok((fid, wal))
+    }
+
+    fn sorted_fids(&self) -> Vec<u32> {
+        let core = self.core.read().unwrap();
+        let mut to_be_deleted = HashSet::new();
+        for fid in core.files_to_delete.iter().cloned() {
+            to_be_deleted.insert(fid);
+        }
+        let mut result = vec![];
+        for (fid, _) in core.files_map.iter() {
+            if to_be_deleted.get(fid).is_none() {
+                result.push(*fid);
+            }
+        }
+        result.sort();
+        result
+    }
+
+    /// Open value log directory
+    pub fn open(&self) -> Result<()> {
```

Comment:
Why not make it a contructor?

Reply:
Now `open` is private, and the vlog directory will be opened upon `new` is called.